watchtower/wtdb/client_db.go

```diff
@@ -0,0 +1,902 @@
+package wtdb
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+const (
+	// clientDBName is the filename of client database.
+	clientDBName = "wtclient.db"
+)
+
+var (
+	// csessionKeyIndexBkt is a top-level bucket storing:
+	//   tower-id -> reserved-session-key-index (uint32).
+	csessionKeyIndexBkt = []byte("client-session-key-index-bucket")
+
+	// cchanSummaryBkt is a top-level bucket storing:
+	//   channel-id -> encoded ClientChanSummary.
+	cchanSummaryBkt = []byte("client-channel-summary-bucket")
+
+	// csessionBkt is a top-level bucket storing:
+	//   session-id => csessionBody -> encoded ClientSessionBody
+	//              => csessionCommits => seqnum -> encoded CommittedUpdate
+	//              => csessionAcks => seqnum -> encoded BackupID
+	csessionBkt = []byte("client-session-bucket")
+
+	// csessionBody is a sub-bucket of csessionBkt storing only the body of
+	// the ClientSession.
+	csessionBody = []byte("client-session-body")
+
+	// csessionBody is a sub-bucket of csessionBkt storing:
+	//    seqnum -> encoded CommittedUpdate.
+	csessionCommits = []byte("client=session-commits")
```

Comment:
Commits are persisted to be able to upload them again on restart. I think the idea is not to lose any commits. But `TowerClient.BackupState` doesn't look to be waiting until the commit has been saved to disk. Is that ok?

Reply:
Related question: to what extend would it be possible to reconstruct those backup tasks from the normal `lnd` channel state db?

Reply:
> Related question: to what extend would it be possible to reconstruct those backup tasks from the normal lnd channel state db?

Given a tower's policy and a revoked commitment transaction, the justice transaction construction is deterministic. So we can recreate tasks for any historical commitment height of any channel. That's primarily the reason it's okay not write them to disk at the time of the handoff, since they can be replayed.

> Commits are persisted to be able to upload them again on restart. I think the idea is not to lose any commits. But TowerClient.BackupState doesn't look to be waiting until the commit has been saved to disk. Is that ok?

Yes we only commit to the updates the moment before sending them to the tower, so that we don't forget which (seqnum, update) we sent the tower, which prevents us from equivocating on a particular sequence number. After restarting, we then replay those `CommittedUpdates` if we didn't get an ack before shutting down.

If the updates are queued, but say the tower is not reachable before the client shuts down, those updates atm will be dropped. However, the plan in the future is to reconcile any gaps that don't appear in `CommittedUpdates` or `AckedUpdates`, and have the channel resubmit the missing ones in the process of registering the channel with the client :)

Reply:
From a db minimalism perspective, could the watchtower be implemented by just keeping the last submitted commitment height and reconstructing everything from the channel?

Reply:
that's pretty close to how it works today, though there is extra persistent state it needs to track to engage in the protocol with the tower